File: cpp/0206_Reverse_Linked_List.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cmath>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *ret = NULL, *ptr = head;
        while (head) {
            ptr = head->next;

            head->next = ret;
            ret = head;

            head = ptr;
        }

        return ret;
    }
};

int main(int argc, char *argv[]) {
    Solution solution;
    ListNode* head;

    // Input: head = [1,2,3,4,5]
    // Output: [5,4,3,2,1]
    head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    head = solution.reverseList(head);
    while(head) {
        cout << head->val << " ";
        head = head->next;
    }
    cout << endl;

    // Input: head = [1,2]
    // Output: [2,1]
    head = new ListNode(1);
    head->next = new ListNode(2);
    head = solution.reverseList(head);
    while(head) {
        cout << head->val << " ";
        head = head->next;
    }
    cout << endl;

    // Input: head = []
    // Output: []
    head = nullptr;
    while(head) {
        cout << head->val << " ";
        head = head->next;
    }
    cout << endl;

    return 0;
}

File: py/0206_Reverse_Linked_List.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        ptr = None
        while head:
            entry = head
            head = head.next
            entry.next = ptr
            ptr = entry
        return ptr

        # recursively
        """
        return self.helper(head, None)
        """

    def helper(self, head: Optional[ListNode], prev: ListNode):
        if not head:
            return prev
        entry = head.next
        head.next = prev
        return self.helper(entry, head)


if __name__ == "__main__":
    solution = Solution()

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    ret = solution.reverseList(head)
    while ret != None:
        print(ret.val)
        ret = ret.next
    print("---")

    head = ListNode(1)
    head.next = ListNode(2)
    ret = solution.reverseList(head)
    while ret != None:
        print(ret.val)
        ret = ret.next
    print("---")

    head = None
    ret = solution.reverseList(head)
    while ret != None:
        print(ret.val)
        ret = ret.next
    print("---")

    sys.exit(0)
